File: zephyr_one/mobile/android/app/src/main/kotlin/one/zephyr/mobile/app/LinkPeerResolver.kt
package one.zephyr.mobile.app

import java.net.InetAddress
import java.net.URI
import java.net.UnknownHostException

/**
 * Turns a Link peer URL into an IP literal plus the original hostname.
 *
 * The embedded Go core is built CGO_ENABLED=0, so it has no Android DNS
 * (cgo resolver) and no /etc/resolv.conf in the app netns. Dialing
 * `https://example.com/...` from Go therefore fails even when OkHttp on the
 * same device can already reach the host. Resolve here with
 * [InetAddress.getAllByName], then tell Go to connect to the IP while keeping
 * the original hostname for TLS SNI and the HTTP Host header.
 */
internal data class LinkPeerTarget(
    val url: String,
    val serverName: String,
)

internal object LinkPeerResolver {

    fun resolve(
        serverUrl: String,
        lookup: (String) -> Array<InetAddress> = { InetAddress.getAllByName(it) },
    ): LinkPeerTarget {
        val uri = URI(serverUrl)
        val host = uri.host?.takeIf { it.isNotBlank() }
            ?: throw IllegalArgumentException("Link URL missing host")
        if (isLiteralIp(host)) return LinkPeerTarget(serverUrl, host)
        val ip = lookup(host).firstOrNull()?.hostAddress
            ?.let(::canonicalIpLiteral)
            ?: throw UnknownHostException(host)
        val encodedIp = if (ip.contains(':')) "[$ip]" else ip
        val authority = if (uri.port != -1) "$encodedIp:${uri.port}" else encodedIp
        val rewritten = URI(uri.scheme, authority, uri.path, uri.query, uri.fragment).toString()
        return LinkPeerTarget(rewritten, host)
    }

    internal fun canonicalIpLiteral(hostAddress: String): String = hostAddress.substringBefore('%')

    internal fun isLiteralIp(host: String): Boolean {
        val value = host.removePrefix("[").removeSuffix("]")
        if (value.contains(':')) return true
        val parts = value.split('.')
        if (parts.size != 4) return false
        return parts.all { part ->
            part.isNotEmpty() && part.all(Char::isDigit) && part.toInt() in 0..255
        }
    }
}

File: zephyr_agent/lib/agent/link_peer_resolver.dart
import 'dart:io';

/// Turns a Link peer URL into IP literals plus the original hostname.
///
/// Mirrors Android LinkPeerResolver: the embedded Go core is CGO_ENABLED=0
/// and has no cgo DNS. Resolve here, then tell Go to connect to each IP
/// while keeping the original hostname for TLS SNI / HTTP Host.
class LinkPeerTarget {
  final String url;
  final String serverName;
  const LinkPeerTarget(this.url, this.serverName);
}

Future<List<LinkPeerTarget>> resolveLinkPeers(String serverUrl) async {
  final uri = Uri.parse(serverUrl);
  final host = uri.host;
  if (host.isEmpty) throw ArgumentError('Link URL missing host');
  if (_isLiteralIp(host)) return [LinkPeerTarget(serverUrl, host)];
  final addresses = await InternetAddress.lookup(host);
  final ips = <String>{};
  for (final addr in addresses) {
    final ip = addr.address.split('%').first;
    if (ip.isNotEmpty) ips.add(ip);
  }
  if (ips.isEmpty) throw SocketException('Failed host lookup: $host');
  final ordered = ips.toList()
    ..sort((a, b) {
      final a6 = a.contains(':');
      final b6 = b.contains(':');
      if (a6 != b6) return a6 ? 1 : -1;
      return a.compareTo(b);
    });
  return ordered.map((ip) => _rewrite(uri, host, ip)).toList();
}

LinkPeerTarget _rewrite(Uri uri, String serverName, String ip) {
  final encoded = ip.contains(':') ? '[$ip]' : ip;
  final rewritten = uri.replace(host: encoded.contains('[') ? ip : ip);
  // Uri.replace with IPv6 needs the host without brackets; Dart adds them.
  return LinkPeerTarget(rewritten.toString(), serverName);
}

bool _isLiteralIp(String host) {
  final value = host.replaceAll('[', '').replaceAll(']', '');
  if (value.contains(':')) return true;
  final parts = value.split('.');
  if (parts.length != 4) return false;
  return parts.every((part) {
    final n = int.tryParse(part);
    return n != null && n >= 0 && n <= 255;
  });
}
